kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/CustomResourceDefinitionContext.java

```diff
@@ -59,40 +58,86 @@ public String getVersion() {
   public String getKind() {
     return kind;
   }
-
-  public static CustomResourceDefinitionContext fromCustomResourceType(Class<? extends HasMetadata> customResource) {
-    HasMetadata instance;
+  
+  public static CustomResourceDefinitionBuilder v1beta1CRDFromCustomResourceType(Class<? extends CustomResource> customResource) {
+    try {
+      final CustomResource instance = customResource.getDeclaredConstructor().newInstance();
+    
+      final String kind = instance.getKind();
+      final String version = instance.getVersion();
+  
+      return new CustomResourceDefinitionBuilder()
+        .withKind(kind)
+        .withNewMetadata()
+        .withName(instance.getCRDName())
+        .endMetadata()
+        .withNewSpec()
+        .withGroup(instance.getGroup())
+        .withVersion(version) // also set version to the first (and only) versions item
+        .addNewVersion().withName(version).endVersion()
+        .withScope(instance.getScope())
+        .withNewNames()
+        .withKind(kind)
+        .withPlural(instance.getPlural())
+        .withSingular(instance.getSingular())
+        .endNames()
+        .endSpec();
+    } catch (InstantiationException | IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
+      throw KubernetesClientException.launderThrowable(e);
+    }
+  }
+  
+  public static io.fabric8.kubernetes.api.model.apiextensions.v1.CustomResourceDefinitionBuilder v1CRDFromCustomResourceType(Class<? extends CustomResource> customResource) {
     try {
-      instance = customResource.newInstance();
-
-      String kind = Utils.isNullOrEmpty(instance.getKind()) ? customResource.getSimpleName() : instance.getKind();
-      String name = kind.toLowerCase();
-      String plural = Pluralize.toPlural(name);
-      String group = ApiVersionUtil.apiGroup(instance, null);
-      String version = ApiVersionUtil.apiVersion(instance, "v1");
-      String scope = instance instanceof Namespaced ? "Namespaced" : "Cluster";
-
-      return new CustomResourceDefinitionContext.Builder()
-        .withGroup(group)
-        .withVersion(version)
-        .withScope(scope)
-        .withName(name)
-        .withPlural(plural)
+      final CustomResource instance = customResource.getDeclaredConstructor().newInstance();
+    
+      String kind = instance.getKind();
+  
+      return new io.fabric8.kubernetes.api.model.apiextensions.v1.CustomResourceDefinitionBuilder()
         .withKind(kind)
+        .withNewMetadata()
+        .withName(instance.getCRDName())
+        .endMetadata()
+        .withNewSpec()
+        .withGroup(instance.getGroup())
+        .addNewVersion().withName(instance.getVersion()).endVersion()
```

Comment:
I think we would also need to set `spec.versions[].schema.openAPIV3Schema`, `spec.versions[].served` and `spec.versions[].storage`. If I don't add these fields in `apiextensions.k8s.io/v1#CustomResourceDefinition` I get these validation errors:
```
The CustomResourceDefinition "crontabs.stable.example.com" is invalid: spec.versions[0].schema.openAPIV3Schema: Required value: schemas are required
```
```
error: error validating "crontab-crd.yml": error validating data: [ValidationError(CustomResourceDefinition.spec.versions[0]): missing required field "served" in io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.CustomResourceDefinitionVersion, ValidationError(CustomResourceDefinition.spec.versions[0]): missing required field "storage" in io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.CustomResourceDefinitionVersion]; if you choose to ignore these errors, turn validation off with --validate=false
```

Reply:
Those will need to be added by the user always.
The idea is to work on the inference of this spec from the POJO, but I would delay this to a future PR.

https://github.com/fabric8io/kubernetes-client/issues/2611#issuecomment-745795616